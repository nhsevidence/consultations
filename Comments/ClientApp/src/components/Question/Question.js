import React, { Component, Fragment } from "react";
import Moment from "react-moment";

import { Answer } from "../Answer/Answer";
//import stringifyObject from "stringify-object";

type QuestionTypeType = {
	description: string,
	hasTextAnswer: boolean,
	hasBooleanAnswer: boolean
};

type QuestionType = {
	questionId: number,
	questionText: string,
	questionTypeId: number,
	questionOrder: number,
	lastModifiedDate: Date,
	lastModifiedByUserId: string,
	questionType: QuestionTypeType,
	answers: Array<AnswerType>,
	commentOn: string
};

type StatusType = {
	statusId: number,
	name: string
};

type AnswerType = {
	answerId: number,
	answerText: string,
	answerBoolean: boolean,
	questionId: number,
	lastModifiedDate: Date,
	lastModifiedByUserId: string,
	statusId: number,
	status: StatusType
};

type PropsType = {
	staticContext?: any,
	saveAnswerHandler: Function,
	deleteAnswerHandler: Function,
	question: QuestionType,
	readOnly: boolean,
	isVisible: boolean, //doesn't appear to get used.
};

type StateType = {
	questionId: number
};

export class Question extends Component<PropsType, StateType> {
	constructor() {
		super();
		this.state = {
			unsavedChanges: false
		};
	}

	isTextSelection = (question) => question.commentOn && question.commentOn.toLowerCase() === "selection" && question.quote;

	render() {
		if (!this.props.question) return null;		

		let answers = this.props.question.answers;
		if (answers === null || answers.length < 1){
			answers = [{
				answerId: -1,
				questionId: this.props.question.questionId
			}];
		}

		return (

			<li className="CommentBox">
				{!this.isTextSelection(this.props.question) &&
				<Fragment>
					<h1 data-qa-sel="comment-box-title" className="CommentBox__title mt--0 mb--d">
						Question on: <span className="text-lowercase">{this.props.question.commentOn}</span>
						<br/>
						{this.props.question.quote}
					</h1>
				</Fragment>
				}

				{this.isTextSelection(this.props.question) &&
				<Fragment>
					<h1 data-qa-sel="comment-box-title" className="CommentBox__title mt--0 mb--d">
						Question on: <span className="text-lowercase">{this.props.question.commentOn}</span>
					</h1>
					<div className="CommentBox__quote mb--d">{this.props.question.quote}</div>
				</Fragment>
				}
				<p><strong>{this.props.question.questionText}</strong></p>
				{answers.map((answer) => {
					return (
						<Answer
							readOnly={this.props.readOnly}
							isVisible={this.props.isVisible}
							key={answer.answerId}
							unique={`Answer${answer.answerId}`}
							answer={answer}
							saveHandler={this.props.saveAnswerHandler}
							deleteHandler={this.props.deleteAnswerHandler}
						/>
					);
				})}

			</li>
		);
	}
}

export default Question;
