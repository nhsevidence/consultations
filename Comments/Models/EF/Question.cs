using System;
using System.Collections.Generic;

namespace Comments.Models
{
    public partial class Question
    {
        private Question()
        {
            Answer = new HashSet<Answer>();
        }

        public int QuestionId { get; set; }
        public int LocationId { get; set; }
        public string QuestionText { get; set; }
        public int QuestionTypeId { get; set; }
        public byte? QuestionOrder { get; set; }

        public Location Location { get; set; }
        public QuestionType QuestionType { get; set; }
        public ICollection<Answer> Answer { get; set; }
    }
}
