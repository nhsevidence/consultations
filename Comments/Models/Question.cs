using System;
using System.Collections.Generic;

namespace Comments.Models
{
    public partial class Question
    {
        public Question(int locationId, string questionText, int questionTypeId, byte? questionOrder, Location location,
            QuestionType questionType, ICollection<Answer> answer)
        {
            LocationId = locationId;
            QuestionText = questionText ?? throw new ArgumentNullException(nameof(questionText));
            QuestionTypeId = questionTypeId;
            QuestionOrder = questionOrder;
            Location = location;
            QuestionType = questionType;
            Answer = answer;
        }

        public void UpdateFromViewModel(ViewModels.Question question)
        {
            LastModifiedByUserId = question.LastModifiedByUserId;
            QuestionText = question.QuestionText ?? throw new ArgumentNullException(nameof(question.QuestionText));
            QuestionOrder = question.QuestionOrder;
            Location.UpdateFromViewModel(question as ViewModels.Location);
        }
    }
}
