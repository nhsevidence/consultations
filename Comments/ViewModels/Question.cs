using System;
using System.Collections.Generic;
using System.Linq;

namespace Comments.ViewModels
{
	public class Question : Location
    {
        public Question() { } //only here for model binding. don't use it in code.
        public Question(Models.Location location, Models.Question question) : base(location.LocationId, location.SourceURI, location.HtmlElementID,
            location.RangeStart, location.RangeStartOffset, location.RangeEnd, location.RangeEndOffset, location.Quote, location.Order, show: true, section: location.Section)
        {
            QuestionId = question.QuestionId;
            QuestionText = question.QuestionText;
            QuestionTypeId = question.QuestionTypeId;
            LastModifiedByUserId = question.LastModifiedByUserId;
            LastModifiedDate = question.LastModifiedDate;

            QuestionType = new QuestionType(question.QuestionType);
            if (!(question.Answer is null))
                Answers = question.Answer.Select(answer => new Answer(answer)).ToList();
        }

        public int QuestionId { get; set; }
        public string QuestionText { get; set; }
        public int QuestionTypeId { get; set; }
        public Guid LastModifiedByUserId { get; set; }
        public DateTime LastModifiedDate { get; set; }

        public ViewModels.QuestionType QuestionType { get; set; }
        public IList<ViewModels.Answer> Answers { get; set; }
    }

    public class QuestionType
    {
        public QuestionType() { } //only here for model binding. don't use it in code.
        public QuestionType(Models.QuestionType questionType)
        {
	        QuestionTypeId = questionType.QuestionTypeId;
			Description = questionType.Description;
            HasTextAnswer = questionType.HasTextAnswer;
            HasBooleanAnswer = questionType.HasBooleanAnswer;
        }
	    public int QuestionTypeId { get; set; }
		public string Description { get; set; }
        public bool HasTextAnswer { get; set; }
        public bool HasBooleanAnswer { get; set; }
    }
}
