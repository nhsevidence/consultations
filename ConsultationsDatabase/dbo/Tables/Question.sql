CREATE TABLE [dbo].[Question] (
    [QuestionID]     INT            IDENTITY (1, 1) NOT NULL,
    [LocationID]     INT            NOT NULL,
    [QuestionText]   NVARCHAR (MAX) NOT NULL,
    [QuestionTypeID] INT            NOT NULL,
    [QuestionOrder]  TINYINT        NULL,
    [CreatedByUserID] UNIQUEIDENTIFIER NOT NULL, 
    [CreatedDate] DATETIME2 NOT NULL DEFAULT GETDATE(), 
    [LastModifiedByUserID] UNIQUEIDENTIFIER NOT NULL, 
    [LastModifiedDate] DATETIME2 NOT NULL, 
    [IsDeleted] BIT NOT NULL DEFAULT 0, 
    CONSTRAINT [PK_Question] PRIMARY KEY CLUSTERED ([QuestionID] ASC),
    CONSTRAINT [FK_Question_Location] FOREIGN KEY ([LocationID]) REFERENCES [dbo].[Location] ([LocationID]),
    CONSTRAINT [FK_Question_QuestionType] FOREIGN KEY ([QuestionTypeID]) REFERENCES [dbo].[QuestionType] ([QuestionTypeID])
);

