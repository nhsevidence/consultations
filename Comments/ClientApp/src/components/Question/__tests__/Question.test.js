import React from "react";
import { mount } from "enzyme";
import { Question } from "../Question";
import questionWithAnswer from "./questionWithAnswer.json";
import questionWithoutAnswer from "./questionWithoutAnswer.json";
import toJson from "enzyme-to-json";

describe("[ClientApp] ", () => {
	describe("Question Component", () => {
		const fakePropsWithAnswer = {
			readOnly: false,
			isVisible: true,
			key: 1,
			unique: "unique",
			question: questionWithAnswer,
			showAnswer: true,
		};

		const fakePropsWithoutAnswer = {
			readOnly: false,
			isVisible: true,
			key: 1,
			unique: "unique",
			question: questionWithoutAnswer,
			showAnswer: false,
		};

		it("should match snapshot with answer", () => {
			const wrapper = mount(
				<Question {...fakePropsWithAnswer} />,					
			);

			expect(toJson(wrapper, {
				noKey: true,
				mode: "deep",
			})).toMatchSnapshot();			
		});

		it("should match snapshot without answer", () => {
			const wrapper = mount(
				<Question {...fakePropsWithoutAnswer} />,					
			);

			expect(toJson(wrapper, {
				noKey: true,
				mode: "deep",
			})).toMatchSnapshot();			
		});
	});
});
