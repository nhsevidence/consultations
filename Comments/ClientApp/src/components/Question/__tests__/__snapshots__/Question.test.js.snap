// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[ClientApp]  Question Component should match snapshot with answer 1`] = `
<li
  className="CommentBox"
>
  <h1
    className="CommentBox__title mt--0 mb--d"
    data-qa-sel="comment-box-title"
  >
    Question on: 
    <span
      className="text-lowercase"
    >
      Consultation
    </span>
    <br />
  </h1>
  <p>
    <strong>
      Which areas will have the biggest impact on practice and be challenging to implement? Please say for whom and why.
    </strong>
  </p>
  <section
    role="form"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form__group form__group--textarea mb--0"
      >
        <textarea
          className="form__input form__input--textarea"
          data-qa-sel="Comment-text-area"
          disabled={false}
          id="Answer1007"
          onChange={[Function]}
          placeholder="Enter your answer here"
          value="some answer text"
        />
      </div>
      <span
        className="ml--0 CommentBox__savedIndicator"
      >
        Saved
      </span>
      <button
        className="btn mr--0 right"
        data-qa-sel="delete-comment-button"
        onClick={[Function]}
      >
        Delete
      </button>
    </form>
  </section>
</li>
`;

exports[`[ClientApp]  Question Component should match snapshot without answer 1`] = `
<li
  className="CommentBox"
>
  <h1
    className="CommentBox__title mt--0 mb--d"
    data-qa-sel="comment-box-title"
  >
    Question on: 
    <span
      className="text-lowercase"
    >
      Consultation
    </span>
    <br />
  </h1>
  <p>
    <strong>
      Which areas will have the biggest impact on practice and be challenging to implement? Please say for whom and why.
    </strong>
  </p>
  <section
    role="form"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form__group form__group--textarea mb--0"
      >
        <textarea
          className="form__input form__input--textarea"
          data-qa-sel="Comment-text-area"
          disabled={false}
          id="Answer-1"
          onChange={[Function]}
          placeholder="Enter your answer here"
        />
      </div>
    </form>
  </section>
</li>
`;
